### PR TITLE
#167313407 Build out commenting on articles REST endpoints(CRUD)

### PR DESCRIPTION
**Pull request by nshutijonathan**
#### What does this PR do?
-Has commenting on the article features
#### Description of Task to be completed?
- Added functionality to comment on articles
- Added get all comment functionality
- Added delete comment functionality
- Added update comment functionality 
- Protected the routes
This PR has the following endpoints working 
 - POST/http://localhost:8000/api/v1/comments/hello-guys-1oqnkmlx2fjze48tc2
- GET/http://localhost:8000/api/v1/comments
- UPDATE/http://localhost:8000/api/v1/comments/</comment id>
- DELETE/http://localhost:8000/api/v1/comments/</comment id>
#### How should this be manually tested?
- Download or clone the tutorial project code from this repo, in your terminal then git checkout `ft2-user-comment-articles-167313407` ,install all required NPM packages by running NPM install from the command line in the project root folder (where the package.json is located).
- Start the API by running NPM start from the command line in the project root folder, you should see the message Server listening on port 8000. You can test the API directly using an application such as Postman or you can test it with one of the single pages in the browser.
- to test authentication and authorization, add this to the header
Key: :x-access-token,value:JWT TOKEN
#### Any background context you want to provide?
 - Implemented the endpoint logic using Sequelize framework 
#### What are the relevant pivotal tracker stories?
[#167313407](https://www.pivotaltracker.com/n/projects/2372121/stories/167313407)
#### Screenshots (if appropriate)
image
image
image
This photo shows how I create thread over the comment, I pass the id of the comment in the req.body then pass also the body and then the threads are created and they are stored in the same database with comments
image

### CODE REVIEW

**Review by nkalyesubula — APPROVED** [review]
LGTM

**Review by nkalyesubula — CHANGES_REQUESTED** [review]
Good so far. Please finish up the threading functanality

**Comment by nshutijonathan** [comment]
> Good so far. Please finish up the threading functionality

Thank You, I was finding out the Implementation logic as I told, I am looking forward to having it done

**Review by salviosage — APPROVED** [review]
LGTM

**Comment by nshutijonathan** [comment]
> Good so far. Please finish up the threading functionality
I have finished implementing the functionality of replying to the comments(threading)

**Review by nkalyesubula — APPROVED** [review]
LGTM

**Review by habinezadalvan — APPROVED** [review]
You've done a great work @nshutijonathan

File: src/controllers/comments.controller.js [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+    const comments = await CommentsDb.findAll();
+    if (!comments) {
+      return res.status(200).send({
+        message: 'No comments found'
+      });
+    }
+    return res.status(200).send({
+      status: 200,
+      message: 'All comments successfully retrieved',
+      comments
```
Comment (minega25) [inline src/controllers/comments.controller.js:116]: @nshutijonathan For the response format we agreed to be in this format:
{
  status: 200,
  message: 'All comments successfully retrieved',
  data: comments
}
However @bdushimi suggested we remove the status attribute. I would like to know which  format our response for success response we finally decided to follow

File: src/models/comment.js [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+      type: DataTypes.INTEGER,
+      allowNull: true
+    },
+    body: DataTypes.STRING
+  }, {});
+  Comment.associate = function(models) {
+    // associations can be defined here
+    Comment.belongsTo(models.user,{
+      foreignKey:'id',as:'commentAuthor'
+    });
```
Comment (minega25) [inline src/models/comment.js:25]: @nshutijonathan I thought the Article model has a one to many relationship with the Comment model. What is your opinion about this?

**Review by minega25 — CHANGES_REQUESTED** [review]
Impressive work @nshutijonathan. Just a few things i mentioned i would like to get more clarification on, in the comments included.

File: src/controllers/comments.controller.js (hunk 1 above)
Comment (nshutijonathan) [reply src/controllers/comments.controller.js:116]: Yes I recall this, but We have concluded to come up the techniques of utils for this upcoming sprint and We shall raise the chore of refactoring all codes and use utils as well, I think you agree with me

File: src/models/comment.js (hunk 2 above)
Comment (nshutijonathan) [reply src/models/comment.js:25]: In my comment's models, I have those relationships  Check very well references: {
        model: 'Articles',
        key: 'slug'
      },
      onUpdate: 'CASCADE',
      onDelete: 'CASCADE'
    },

Comment (minega25) [reply src/models/comment.js:25]: Cool. Thank you for the clarification

File: src/controllers/comments.controller.js (hunk 1 above)
Comment (minega25) [reply src/controllers/comments.controller.js:116]: Oh okay. Cool

**Review by minega25 — APPROVED** [review]

**Pull request merged**